nsdb-cluster/src/main/scala/io/radicalbit/nsdb/cluster/actor/AbstractClusterListener.scala

```diff
@@ -0,0 +1,250 @@
+                      case LocationWithCoordinates(_, _, location) => location
+                    })
+                }
+              }
+              .map(ErrorManagementUtils.partitionResponses[LocationsAdded, AddLocationsFailed])
+              .retry(delay, retries) {
+                case (_, addLocationsFailedList) => addLocationsFailedList.isEmpty
+              }
+              .onComplete {
+                case Success((_, failures)) if failures.isEmpty =>
+                  val nodeName = createNodeName(member)
+                  mediator ! Subscribe(NODE_GUARDIANS_TOPIC, nodeActorsGuardian)
+                  mediator ! Publish(NSDB_LISTENERS_TOPIC, NodeAlive(nodeId, nodeName))
+                  NSDbClusterSnapshot(context.system).addNode(nodeId, nodeName)
+                  onSuccessBehaviour(readCoordinator, writeCoordinator, metadataCoordinator, publisherActor)
+                case e =>
+                  onFailureBehaviour(member, e)
+              }
+          case unknownResponse =>
+            log.error(s"unknown response from nodeActorsGuardian ? GetNodeChildActors $unknownResponse")
+        }
+    case NodeAlive(nodeId, address) =>
+      NSDbClusterSnapshot(context.system).addNode(nodeId, address)
+    case UnreachableMember(member) =>
+      log.info("Member detected as unreachable: {}", member)
+
+      val nodeName = createNodeName(member)
+
+      unsubscribeNode(nodeName)
+
+      NSDbClusterSnapshot(context.system).removeNode(nodeName)
+
+    case MemberRemoved(member, previousStatus) =>
+      log.info("Member is Removed: {} after {}", member.address, previousStatus)
+
+      val nodeName = createNodeName(member)
+
+      unsubscribeNode(nodeName)
+
+      NSDbClusterSnapshot(context.system).removeNode(nodeName)
+    case _: MemberEvent => // ignore
+    case DiskOccupationChanged(nodeName, usableSpace, totalSpace) =>
+      log.debug(s"received usableSpace $usableSpace and totalSpace $totalSpace for nodeName $nodeName")
+      nsdbMetrics.put(
+        nodeName,
+        Set(
+          NodeMetrics(
+            createAddress(nodeName),
+            System.currentTimeMillis(),
+            Set(Metric(NSDbMetrics.DiskFreeSpace, usableSpace, None),
+                Metric(NSDbMetrics.DiskTotalSpace, totalSpace, None))
+          ))
+      )
+      log.debug(s"nsdb metrics $nsdbMetrics")
+    case ClusterMetricsChanged(nodeMetrics) =>
+      log.debug(s"received metrics $nodeMetrics")
+      akkaClusterMetrics = nodeMetrics.groupBy(nodeMetric => createNodeName(nodeMetric.address))
+      Try {
+        val fs = Files.getFileStore(Paths.get(indexPath))
+        mediator ! Publish(NSDB_METRICS_TOPIC, DiskOccupationChanged(selfNodeName, fs.getUsableSpace, fs.getTotalSpace))
+        log.debug(s"akka cluster metrics $akkaClusterMetrics")
+      }.recover {
+        // if the fs path has not been created yet, the occupation ratio will be 100.0
```

Comment:
may this imply eventual orchestrator might spawn new replicas indefinitely?

Reply:
This part of the codebase only calculates and collects the metrics that are used to pick up a node.
The disk occupation metric is calculated as 
```freeSpace.toDouble / totalSpace```

the value varies from 0.0 to 1.0; the first boundary means no capacity and the latter full capacity.
This recovery branch is reached when the data directory hasn't been created yet, so it's safe to imply that the disk is fully available, so the trick that has been used  is to set the same value for usable space and total space in order to get 1.0 (full capacity) as the disk occupation metric.

The comment above is in fact misleading.  I'm updating it.